File: alert_page.py
from flask import Flask, request

app = Flask(__name__)
port = 3200

# Array to store alerts
alerts = []

@app.route('/')
def home():
    return 'Webhook server is running'

@app.route('/grafana-alert', methods=['POST'])
def grafana_alert():
    alert = request.json
    print('Received alert:', alert)

    # Store the alert in the alerts array
    alerts.append(alert)

    return 'Alert received', 200

@app.route('/alerts', methods=['GET'])
def get_alerts():
    alert_count = len(alerts)
    alert_details = f"""
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Alert Dashboard</title>
            <style>
                body {{
                    font-family: 'Arial', sans-serif;
                    margin: 0;
                    padding: 0;
                    background-color: #f8f9fa;
                    color: #333;
                }}
                header {{
                    background-color: #d90429;
                    color: white;
                    padding: 20px;
                    text-align: center;
                }}
                .container {{
                    max-width: 800px;
                    margin: 30px auto;
                    background: white;
                    border-radius: 8px;
                    box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
                    padding: 20px;
                }}
                h1 {{
                    font-size: 24px;
                    color: white;
                }}
                .alert-count {{
                    font-size: 18px;
                    color: #d90429;
                    margin-bottom: 20px;
                }}
                .alert-card {{
                    border: 1px solid #d90429;
                    background: #ffe3e3;
                    padding: 15px;
                    margin-bottom: 15px;
                    border-radius: 5px;
                }}
                .alert-card p {{
                    margin: 0;
                    font-weight: bold;
                    color: #d90429;
                }}
                .no-alerts {{
                    text-align: center;
                    font-size: 18px;
                    color: #6c757d;
                }}
                footer {{
                    text-align: center;
                    padding: 10px;
                    font-size: 14px;
                    background-color: #f8f9fa;
                    color: #6c757d;
                    border-top: 1px solid #dee2e6;
                }}
            </style>
        </head>
        <body>
            <header>
                <h1>Alert Dashboard</h1>
            </header>
            <div class="container">
                <p class="alert-count"><strong>Total Alerts:</strong> {alert_count}</p>
    """

    if alert_count > 0:
        for index, alert in enumerate(alerts):
            # Extract the alert name from the message body if available
            message_body = alert.get('message', '')  # Assume message body is in the 'message' key
            alert_name = "Unnamed Alert"
            
            # Attempt to extract "alertname" from the message body
            if "alertname =" in message_body:
                try:
                    alert_name = message_body.split("alertname =")[1].split("-")[0].strip()
                except IndexError:
                    pass  # Keep default name if parsing fails

            alert_details += f"""
                <div class="alert-card">
                    <p>Alert {index + 1}: {alert_name}</p>
                </div>
            """
    else:
        alert_details += "<p class='no-alerts'>No alerts received yet.</p>"

    alert_details += """
            </div>
        </body>
        </html>
    """
    return alert_details

@app.route('/grafana-alert', methods=['GET'])
def grafana_alert_get():
    return 'GET request to /grafana-alert is not supported. This endpoint accepts POST requests only.', 405

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=port)

File: src/alert_page.php
<?php
session_start();

// Initialize alerts array in session
if (!isset($_SESSION['alerts'])) {
    $_SESSION['alerts'] = [];
}

// Handle POST request to store alerts
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $input = file_get_contents('php://input');
    $alert = json_decode($input, true);

    if ($alert) {
        $_SESSION['alerts'][] = $alert;
        http_response_code(200);
        echo "Alert received";
        exit;
    } else {
        http_response_code(400);
        echo "Invalid alert format";
        exit;
    }
}

// Display alerts
if ($_SERVER['REQUEST_METHOD'] === 'GET') {
    $alerts = $_SESSION['alerts'];
    $alertCount = count($alerts);

    echo "<!DOCTYPE html>
    <html lang='en'>
    <head>
        <meta charset='UTF-8'>
        <meta name='viewport' content='width=device-width, initial-scale=1.0'>
        <title>Grafana Alerts</title>
        <style>
            body {
                font-family: Arial, sans-serif;
                margin: 0;
                padding: 0;
                background-color: #f4f4f9;
                color: #333;
            }
            header {
                background-color: #007bff;
                color: white;
                padding: 20px;
                text-align: center;
            }
            .container {
                max-width: 800px;
                margin: 20px auto;
                padding: 20px;
                background: white;
                border-radius: 8px;
                box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
            }
            h1 {
                color: #007bff;
            }
            ul {
                list-style-type: none;
                padding: 0;
            }
            li {
                margin-bottom: 15px;
                border-bottom: 1px solid #ddd;
                padding-bottom: 10px;
            }
            strong {
                color: #007bff;
            }
            .alert-card {
                background: #f9f9fb;
                border: 1px solid #ddd;
                padding: 15px;
                margin-bottom: 10px;
                border-radius: 8px;
            }
            .alert-card ul {
                padding-left: 20px;
                list-style: square;
            }
            a {
                color: #007bff;
                text-decoration: none;
            }
            a:hover {
                text-decoration: underline;
            }
            .no-alerts {
                text-align: center;
                color: #666;
                font-size: 18px;
            }
            footer {
                text-align: center;
                padding: 10px;
                margin-top: 20px;
                font-size: 14px;
                background-color: #f1f1f1;
            }
        </style>
    </head>
    <body>
        <header>
            <h1>Grafana Alerts</h1>
        </header>
        <div class='container'>
            <p><strong>Total Alerts:</strong> $alertCount</p>";

    if ($alertCount > 0) {
        foreach ($alerts as $index => $alert) {
            $labels = isset($alert['labels']) ? $alert['labels'] : [];
            $annotations = isset($alert['annotations']) ? $alert['annotations'] : [];
            
            $labelDetails = "";
            foreach ($labels as $key => $value) {
                $labelDetails .= "<li>$key: $value</li>";
            }

            $annotationDetails = "";
            foreach ($annotations as $key => $value) {
                $annotationDetails .= "<li>$key: $value</li>";
            }

            $status = isset($alert['status']) ? $alert['status'] : 'Unknown';
            $message = isset($alert['message']) ? $alert['message'] : 'No message provided';
            $value = isset($alert['value']) ? $alert['value'] : 'N/A';
            $generatorURL = isset($alert['generatorURL']) ? $alert['generatorURL'] : '#';

            echo "<div class='alert-card'>
                    <p><strong>Alert " . ($index + 1) . "</strong></p>
                    <ul>
                        <li><strong>Status:</strong> $status</li>
                        <li><strong>Message:</strong> $message</li>
                        <li><strong>Value:</strong> $value</li>
                        <li><strong>Labels:</strong>
                            <ul>$labelDetails</ul>
                        </li>
                        <li><strong>Annotations:</strong>
                            <ul>$annotationDetails</ul>
                        </li>
                        <li><strong>Alert URL:</strong> <a href='$generatorURL' target='_blank'>View Alert</a></li>
                    </ul>
                  </div>";
        }
    } else {
        echo "<p class='no-alerts'>No alerts received yet.</p>";
    }

    echo "</div>
    </body>
    </html>";
    exit;
}

// For unsupported methods
http_response_code(405);
echo "Method Not Allowed. Only GET and POST requests are supported.";
exit;